File: leecode/code/105/105.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {


    public int findnum(int[] nums, int start, int end, int target) {
        for (; start <= end; start++) {
            if (nums[start] == target) {
                return start;
            }
        }
        return start;
    }


    public TreeNode helper(int[] preorder, int[] inorder, int pl, int pr, int ol, int or) {
        if (pl > pr || ol > or) {
            return null;
        }


        TreeNode root = new TreeNode(preorder[pl]);
        int mid = findnum(inorder, ol, or, preorder[pl]);
            root.left = helper(preorder, inorder, pl + 1, pl + mid - ol, ol, mid - 1);
            root.right = helper(preorder, inorder, pl + mid - ol + 1, pr, mid + 1, or);
        return root;
    }


    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if (preorder.length == 0) {
            return null;
        }
        TreeNode root = helper(preorder, inorder,
                                 0, preorder.length - 1, 0, inorder.length - 1);
       
        return root;
    }
}

File: leecode/code/105/105.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        if len(preorder) == 0:
            return None


        k = TreeNode(preorder[0])
        mid = inorder.index(preorder[0])


        k.left = self.buildTree(preorder[1:mid + 1], inorder[0:mid])
        k.right = self.buildTree(preorder[mid + 1:], inorder[mid + 1:])


        return k